cmds/ls/fileinfo_freebsd.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	humanize "github.com/dustin/go-humanize"
+)
+
+// From Linux header: /include/uapi/linux/kdev_t.h
+const (
+	minorBits = 8
+	minorMask = (1 << minorBits) - 1
+)
+
+// Matches characters which would interfere with ls's formatting.
+var unprintableRe = regexp.MustCompile("[[:cntrl:]\n]")
+
+// Since `os.FileInfo` is an interface, it is difficult to tweak some of its
+// internal values. For example, replacing the starting directory with a dot.
+// `extractImportantParts` populates our own struct which we can modify at will
+// before printing.
+type fileInfo struct {
+	name         string
+	mode         os.FileMode
+	major, minor uint64
+	uid, gid     uint32
+	size         int64
+	modTime      time.Time
+	symlink      string
+}
+
+func extractImportantParts(n string, fi os.FileInfo) fileInfo {
+	var link string
+
+	s := fi.Sys().(*syscall.Stat_t)
+	if fi.Mode()&os.ModeType == os.ModeSymlink {
+		if l, err := os.Readlink(n); err != nil {
```

Comment:
for things like this our style is usually
var err error
if link, err = osReadlink(n); err != nil {
link = err.Error()
}
and you can skip the else and it's a bit less convoluted.

Reply:
Left alone original code. Now in fileinfo_unix.go